base/log_level.go

```diff
@@ -0,0 +1,60 @@
+package base
+
+import (
+	"errors"
+	"fmt"
+	"sync/atomic"
+)
+
+// LogLevel is used to represent a log level.
+type LogLevel uint32
+
+const (
+	// LEVEL_NONE disables all logging
+	LEVEL_NONE LogLevel = iota
+	// LEVEL_ERROR enables only error logging.
+	LEVEL_ERROR
+	// LEVEL_WARN enables warn, and error logging.
+	LEVEL_WARN
+	// LEVEL_INFO enables info, warn, and error logging.
+	LEVEL_INFO
+	// LEVEL_DEBUG enables all logging.
+	LEVEL_DEBUG
+)
+
+var logLevelNames = []string{"none", "error", "warn", "info", "debug"}
+
+func (l *LogLevel) Set(newLevel LogLevel) {
+	atomic.StoreUint32((*uint32)(l), uint32(newLevel))
+}
+
+// Enabled returns true if the log level is enabled.
+func (l *LogLevel) Enabled(logLevel LogLevel) bool {
+	if l == nil {
+		return false
+	}
+	return atomic.LoadUint32((*uint32)(l)) >= uint32(logLevel)
+}
```

Comment:
I'm not completely convinced that the type alias should be managing the locking on itself.  It feels a bit counter-intuitive; you'd expect the consumer using the alias to have a better understanding of its' concurrency requirements.

Is this primarily in order to support atomic operations during text marshal/unmarshal?